scipy/optimize/tests/test__differential_evolution.py

```diff
@@ -554,3 +559,576 @@ def test_converged(self):
+                    1 - x[1] + (x[2]-4)**2]
+
+        N = NonlinearConstraint(c1, -np.inf, 0)
+        bounds = [(0, 10)]*2
+        constraints = (N)
+
+        res = differential_evolution(f, bounds, strategy='rand1bin', seed=1234,
+                                     constraints=constraints)
+
+        x_opt = (1.22067691, 4.24096915)
+        f_opt = -0.100738
+
+        assert_allclose(f(x_opt), f_opt, atol=2e-7)
+        assert_allclose(res.x, x_opt, atol=5e-3)
+        assert_allclose(res.fun, f_opt, atol=1e-4)
+        assert res.success
+        assert_(np.all(np.array(c1(res.x)) <= 0))
+        assert_(np.all(res.x >= np.array(bounds)[:, 0]))
+        assert_(np.all(res.x <= np.array(bounds)[:, 1]))
+
+    def test_L6(self):
+        # Lampinen ([5]) test problem 6
+        def f(x):
+            x = np.hstack(([0], x))  # 1-indexed to match reference
+            fun = (x[1]-10)**3 + (x[2] - 20)**3
+            return fun
+
+        def c1(x):
+            x = np.hstack(([0], x))  # 1-indexed to match reference
+            return [(x[1]-5)**2 + (x[2] - 5)**2 - 100,
+                    -(x[1]-6)**2 - (x[2] - 5)**2 + 82.81]
+
+        N = NonlinearConstraint(c1, 0, np.inf)
+        bounds = [(13, 100), (0, 100)]
+        constraints = (N)
+        res = differential_evolution(f, bounds, strategy='rand1bin', seed=1234,
+                                     constraints=constraints)
+        x_opt = (14.095, 0.84296)
+        f_opt = -6961.814744
+
+        assert_allclose(f(x_opt), f_opt, atol=1e-6)
+        assert_allclose(res.fun, f_opt, atol=0.05)
+        assert_allclose(res.x, x_opt, atol=1e-4)
+        assert res.success
+        assert_(np.all(np.array(c1(res.x)) >= 0))
+        assert_(np.all(res.x >= np.array(bounds)[:, 0]))
+        assert_(np.all(res.x <= np.array(bounds)[:, 1]))
+
+    def test_L7(self):
+        # Lampinen ([5]) test problem 7
+        def f(x):
+            x = np.hstack(([0], x))  # 1-indexed to match reference
+            fun = (5.3578547*x[3]**2 + 0.8356891*x[1]*x[5] +
+                   37.293239*x[1] - 40792.141)
+            return fun
+
+        def c1(x):
+            x = np.hstack(([0], x))  # 1-indexed to match reference
+            return [
+                    85.334407 + 0.0056858*x[2]*x[5] + 0.0006262*x[1]*x[4] -
+                    0.0022053*x[3]*x[5],
+
+                    80.51249 + 0.0071317*x[2]*x[5] + 0.0029955*x[1]*x[2] +
+                    0.0021813*x[3]**2,
+
+                    9.300961 + 0.0047026*x[3]*x[5] + 0.0012547*x[1]*x[3] +
+                    0.0019085*x[3]*x[4]
+                    ]
+
+        N = NonlinearConstraint(c1, [0, 90, 20], [92, 110, 25])
+
+        bounds = [(78, 102), (33, 45)]+[(27, 45)]*3
+        constraints = (N)
+
+        res = differential_evolution(f, bounds, strategy='rand1bin', seed=1234,
+                                     constraints=constraints)
+
+        x_opt = (78.0, 33.0, 29.995, 45.0, 36.776)
+        f_opt = -30665.608768
```

Comment:
Strange - the value reported in Lampinen is much lower than this one, which was taken from Koziel. But Koziel reports this value here `-30665.608768` as "the optimum", whereas he is careful to note others as "the best known value". I wonder if Lampinen made a mistake when implementing this one.

Reply:
DE is getting ~ -30665.608768 so there's nothing to be changed about the test. Just weird that Lampinen reports a different value.

Reply:
For this particular example Koziel actually reports -30665.5. However, f(x_opt) is -30665.608768 - his evaluated function isn't correct.

Having said that I can't get to the actual minimum, the closest I can get is -30665.538, when `res.x` is within `atol=1e-3` of `x_opt`.

Investigating this further I get `[92.000043492964, 98.8405108381325, 19.999935068412004]` when evaluating the NonlinearConstraint `N.fun(x_opt)`. The first and last values are outside the constraint bounds `[(0, 92), (90, 110), (20, 25)]`. Thus, the reported `x_opt` in Koziel doesn't satisfy the constraints. I wonder if he was using single precision or something.

I'm going to fix this by using our values, coz I trust ours more.